### PR TITLE
GH14: Auto Detect Cake Tasks in VSCode

### PR DESCRIPTION
**Pull request by gep13**
@cake-build/cake-team can you please all have a look at this, and let me know what you think?

### CODE REVIEW

**Comment by gep13** [comment]
In order to run this...

* it is probably easier to remove the existing Cake extension that you will have installed
* check out the latest source code
* open project in VSCode
* press F5
* open a folder that contains some *.cake files
* Press F1 and type `task`
* Select `Tasks: Run Task`

The list of Cake Tasks should then be displayed.

File: src/cakeMain.ts [hunk 1]
```diff
@@ -1,19 +1,101 @@
...
     }));
     // Register the configuration command.
-    context.subscriptions.push(commands.registerCommand('cake.configuration', async () => {
+    context.subscriptions.push(vscode.commands.registerCommand('cake.configuration', async () => {
         installCakeConfiguration();
     }));
+
+    function onConfigurationChanged() {
+        let autoDetect = vscode.workspace.getConfiguration('cake').get<AutoDetect>('autoDetect');
+		if (taskProvider  && autoDetect === 'off') {
```
Comment (mholo65) [inline src/cakeMain.ts]: Indentation?

**Comment by gep13** [comment]
@mholo65 If you could take it for a spin as well on your machine, to make sure that it isn't a case of "works on my machine" I would appreciate it. Thank you!

**Comment by mholo65** [comment]
@gep13 will do when near computer :wink:

File: src/cakeMain.ts (hunk 1 above)
Comment (gep13) [inline src/cakeMain.ts]: @mholo65 not sure on what you think it wrong here? It looks right to me.

Comment (gep13) [inline src/cakeMain.ts]: @mholo65 hmm, hold that thought...

File: build.cake [hunk 2]
```diff
@@ -60,7 +60,7 @@ Task("Npm-Install")
     NpmInstall();
 });
 
-Task("Install-TypeScript)
+Task("Install-TypeScript")
```
Comment (mholo65) [inline build.cake]: Need intellisense?

Comment (gep13) [inline build.cake]: HA HA! Actually, the issue was clearly identifiable once I re-installed the Cake extension (with it's syntax highlighting) 😄

**Comment by mholo65** [comment]
Verified on my computer with two different projects. Works like a charm!

**Comment by gep13** [comment]
>@mholo65 said...
Verified on my computer with two different projects. Works like a charm!

Woot!

Was that a Windows machine?  Do you have access to a Mac or Linux box to test Task execution works as expected?

**Comment by mholo65** [comment]
> @gep13 said...
> Was that a Windows machine? Do you have access to a Mac or Linux box to test Task execution works as expected?

Yes it was. No don't have access to either Linux or Mac. No need for Linux box anymore since Win10 and WSL. Maybe @redth want to try as he was originally interested in #26?

**Comment by gep13** [comment]
Good thinking!  Will wait and see if @redth can also provide some input here.

**Comment by Redth** [comment]
Going to give this a spin...

**Comment by Redth** [comment]
Finally got this going... I have one suggestion.  Let's make the regex pattern less rigid.

For instance, usually my old mono code habits lead me to declaring a task like this:

```
Task ("libs").Does (() => { /* ... */ });
```
Notice the space after `Task` and before `(`.

In theory, it's possible to have whitespace in between parenthesis and quotes, so how about we use this as the regex?

```
/Task\s?\(\s?"(.*?)"\s?\)/g
```

Thoughts?

**Comment by gep13** [comment]
@redth yes, that makes a lot of sense.

How about we go even further and add the regex as a configuration option, and set a default value.

**Comment by Redth** [comment]
Great!

**Comment by gep13** [comment]
>@Redth said...
Great!

I will make the configuration additions, and then I might get you to take another look if that is ok.

**Comment by Redth** [comment]
Just let me know!

**Comment by mholo65** [comment]
Correct me if I'm wrong, but wont the regex proposed by @redth mark zero or one whitespace. I think should throw in a star there to also match many whitespaces. Keep the question mark to make it ungreedy.

I.e whitespace regex should look like this `\s*?`

**Comment by gep13** [comment]
>@mholo65 said...
I.e whitespace regex should look like this \s*?

Yes, I had had the same thought, but hadn't replied about it.  Yes, I believe you are correct.  Just done a quick test here:

https://regex101.com/

And your suggestion of `\s*?` is required.

**Comment by Redth** [comment]
Good catch, for some reason i was thinking `\s` would match one or more which clearly is not the case.

**Comment by mholo65** [comment]
@Redth yeah, `\s` is just whitespace. `\s+` is one or more whitespace. `\s*` is zero to many. Adding the `?` will make the match nongreedy, i.e. will stop at first match.

**Comment by Redth** [comment]
@mholo65 yep, completely obvious after the fact ;) Thanks for noticing that one!

**Comment by gep13** [comment]
@mholo65 @Redth if either of you have a minute, can you have another look at this PR?  You should now have the following configuration options:

image

Which allow configuration of how the Tasks are detected.  The defaults should work for you as expected, but if you need something more the options are there.

**Comment by Redth** [comment]
Looks good to me!

**Comment by patriksvensson** [comment]
@gep13 Would it be possible to group the settings like we've done with the bootstrappers? It makes it easier to see what settings are related. If not, could we add something to the setting namespace to group them?

```json
"cake.taskRunner": {
  "autoDetect": "on",
  "scriptsGlobPattern": "**/*.cake",
  "scriptsExcludePattern": "**/format_C.cake",
  "taskRegularExpression": "lol"
}
```

or

```json
"cake.taskRunner.autoDetect": "on",
"cake.taskRunner.scriptsGlobPattern": "**/*.cake",
"cake.taskRunner.scriptsExcludePattern": "**/format_C.cake",
"cake.taskRunner.taskRegularExpression": "lol"
```

**Comment by gep13** [comment]
@patriksvensson hmmm, it might be, but I will have to check.  The autodetect parameter is an enum, and I am not sure if you can next it inside something else.  But I will give it a try.

Agree on the renaming for consistency regardless.

@agc93 @patriksvensson were either of you actually able to take this for a spin?  Did it work on your machines?  Would love if someone could test this on a Mac.

**Comment by gep13** [comment]
@patriksvensson I have updated the extension to use the following:

image

Thoughts?

**Comment by patriksvensson** [comment]
@gep13 I think it looks awesome 👍

**Comment by patriksvensson** [comment]
@gep13 I can probably test it on my mac tonight if VSCode still runs on it (old OS).

**Pull request merged**